Sources/NIO/Selector.swift

```diff
@@ -406,22 +449,36 @@ final class Selector<R: Registration> {
 
         for i in 0..<ready {
             let ev = events[i]
+            var selectorEvent: SelectorEventSet = ._none
             switch Int32(ev.filter) {
             case EVFILT_USER:
                 // woken-up by the user, just ignore
-                break
+                continue
             case EVFILT_READ:
-                if let registration = registrations[Int(ev.ident)] {
-                    try body((SelectorEvent(io: .read, registration: registration)))
+                selectorEvent.formUnion(.read)
+                fallthrough
+            case EVFILT_EXCEPT:
+                if Int32(ev.flags) & EV_EOF != 0 {
+                    selectorEvent.formUnion(.readEOF)
                 }
-            case EVFILT_WRITE:
-                if let registration = registrations[Int(ev.ident)] {
-                    try body((SelectorEvent(io: .write, registration: registration)))
+                if ev.fflags != 0 {
+                    selectorEvent.formUnion(.reset)
                 }
+            case EVFILT_WRITE:
+                selectorEvent.formUnion(.write)
             default:
                 // We only use EVFILT_USER, EVFILT_READ and EVFILT_WRITE.
                 fatalError("unexpected filter \(ev.filter)")
             }
+            if selectorEvent != ._none {
```

Comment:
Might be nicer to express this as a `guard`: `guard selectorEvent != .none, let registration = registrations[Int(ev.ident)] else { continue }`